mrp_bom_hierarchy/model/mrp.py

```diff
@@ -0,0 +1,267 @@
+
+    def _bom_hierarchy_indent_calc(self, cr, uid, ids, prop, unknow_none,
+                                   unknow_dict):
+        if not ids:
+            return []
+        res = []
+        for bom in self.browse(cr, uid, ids, context=None):
+            data = []
+            b = bom
+            while b:
+                if b.name and b.bom_id:
+                    data.insert(0, '>')
+                else:
+                    data.insert(0, '')
+
+                b = b.bom_id
+            data = ''.join(data)
+            res.append((bom.id, data))
+        return dict(res)
+
+    def _complete_bom_hierarchy_code_calc(
+            self, cr, uid, ids, prop, unknow_none, unknow_dict):
+        if not ids:
+            return []
+        res = []
+        for bom in self.browse(cr, uid, ids, context=None):
+            data = []
+            b = bom
+            while b:
+                if b.code:
+                    data.insert(0, b.code)
+                elif b.position:
+                    data.insert(0, b.position)
+                elif b.product_id.default_code:
+                    data.insert(0, b.product_id.default_code)
+                else:
+                    data.insert(0, '')
+
+                b = b.bom_id
+            data = ' / '.join(data)
+            data = '[' + data + '] '
+
+            res.append((bom.id, data))
+        return dict(res)
+
+    def _complete_bom_hierarchy_name_calc(
+            self, cr, uid, ids, prop, unknow_none, unknow_dict):
+        if not ids:
+            return []
+        res = []
+        for bom in self.browse(cr, uid, ids, context=None):
+            data = []
+            b = bom
+            while b:
+                if b.name:
+                    data.insert(0, b.name)
+                elif b.product_id.name:
+                    data.insert(0, b.product_id.name)
+                else:
+                    data.insert(0, '')
+
+                b = b.bom_id
+
+            data = ' / '.join(data)
+            res.append((bom.id, data))
+        return dict(res)
+
+    def _is_parent(self, cr, uid, ids, prop, unknow_none, unknow_dict):
+        res = {}
+        for bom in self.browse(cr, uid, ids, context=None):
+            if not bom.bom_id:
+                res[bom.id] = True
+            else:
+                res[bom.id] = False
+        return res
+
+    def _product_has_own_bom(self, cr, uid, ids, prop, unknow_none,
+                             unknow_dict):
+        res = {}
+        for bom in self.browse(cr, uid, ids, context=None):
+            bom_ids = self.pool.get('mrp.bom').search(
+                cr, uid, [('product_id', '=', bom.product_id.id),
+                          ('bom_id', '=', False)], context=None)
+            if bom_ids:
+                res[bom.id] = True
+            else:
+                res[bom.id] = False
+        return res
+
+    _columns = {
+        'is_parent': fields.function(_is_parent, string="Is parent BOM",
```

Comment:
You don't need this computed field to make visible/invisible a field. Just put `attrs={'invisible': [('bom_id', '!=', False)]}`